### PR TITLE
Allow deep mask json and custom JSON serializer. Fix non UTF-8 for js…

### PR DESCRIPTION
**Pull request by Meat-Chopper**
Hi,
I see the implementation of JSON data masking in the "develop" branch. But since it has not yet been merged with the master, I have decided to try with my own implementation.
Also there is a bug when JSON is unable to dump non UTF-8 symbols.

### CODE REVIEW

File: spec/lib/http_log_spec.rb [hunk 1]
```diff
@@ -304,6 +310,88 @@ def configure
...
+        before { adapter.send_post_request }
+
+        if adapter_class.method_defined? :send_post_request
+          it { expect(json['request_headers'].to_s).not_to include(secret) }
+          it { expect(json['request_body'].to_s).to include(secret) }
+          it { expect(json['response_body'].to_s).not_to include(secret) }
+        end
+      end
+
+      describe 'Non UTF-8 with JSON log' do
```
Comment (Meat-Chopper) [inline spec/lib/http_log_spec.rb]: It fails in version 1.3.2

File: lib/httplog/configuration.rb [hunk 2]
```diff
@@ -19,35 +19,41 @@ class Configuration
...
+      @log_response            = true
+      @log_benchmark           = true
+      @url_whitelist_pattern   = nil
+      @url_blacklist_pattern   = nil
+      @url_masked_body_pattern = nil
+      @color                   = false
+      @prefix_data_lines       = false
+      @prefix_response_lines   = false
+      @prefix_line_numbers     = false
+      @mask_json               = false
```
Comment (trusche) [inline lib/httplog/configuration.rb]: This parameter seems redundant. One could just enable global JSON masking by setting the URL parameter to `/.*/`. The regex is checked every time anyway.

File: lib/httplog/http_log.rb [hunk 3]
```diff
@@ -52,10 +53,15 @@ def url_approved?(url)
       !config.url_whitelist_pattern || url.to_s.match(config.url_whitelist_pattern)
     end
 
+    def masked_body_url?(url)
+      config.filter_parameters.any? &&
+        (config.url_masked_body_pattern && url.to_s.match(config.url_masked_body_pattern) || config.mask_json)
```
Comment (trusche) [inline lib/httplog/http_log.rb]: See above - the final condition wouldn't be necessary if `url_masked_body_pattern == /.*/`

File: lib/httplog/http_log.rb [hunk 4]
```diff
@@ -127,14 +133,25 @@ def parse_body(body, encoding, content_type)
...
 
-      utf_encoded(body.to_s, content_type)
+      result = utf_encoded(body.to_s, content_type)
+
+      if mask_body && body && !body.empty?
+        if content_type =~ /json/
+          begin
+            result = masked_data(config.json_parser.load(result))
+          rescue => e
+            e.message + ': ' + result
```
Comment (trusche) [inline lib/httplog/http_log.rb]: This silently swallow a failure. Perhaps we should just log an error message and return an empty string so we're not accidentally leaking sensitive data into the logfile. Or looking further down, perhaps that's what's indented already and you forgot to re-raise the exception here?

File: lib/httplog/http_log.rb [hunk 5]
```diff
@@ -236,6 +267,38 @@ def masked(msg, key=nil)
...
     end
 
+    def parse_request(options)
+      return if options[:request_body].nil?
+
+      request = options[:request_headers].find_all do |k, _|
+        # Capitalized for ::HTTP
+        %w[content-type Content-Type content-encoding Content-Encoding].include? k
+      end.to_h.each_with_object({}) { |(k, v), h| h[k.downcase] = v }
+
```
[the remark this answers is not in the document]
Comment (trusche) [reply lib/httplog/http_log.rb]: Firstly, this reads a little hard. I'd replace `k` with `header` on the first block - I tend to reserve single character variable names for single line blocks (as in 276).

[the remark this answers is not in the document]
Comment (trusche) [reply lib/httplog/http_log.rb]: Secondly, it's not at all clear to me what this block does.  Can you add a comment, or perhaps use clearer parameter names? The result of this block is not the request, it's just a bunch of headers. 

But leaving the logic here is fine, to deal with differently capitalized headers in one place. It just needs to be clearer that that's what we're doing here. Perhaps move it to a dedicated class method.

File: spec/lib/http_log_spec.rb [hunk 6]
```diff
@@ -35,7 +35,10 @@
   let(:compact_log)           { HttpLog.configuration.compact_log }
   let(:url_blacklist_pattern) { HttpLog.configuration.url_blacklist_pattern }
   let(:url_whitelist_pattern) { HttpLog.configuration.url_whitelist_pattern }
+  let(:mask_json)             { HttpLog.configuration.mask_json }
+  let(:json_parser)           { HttpLog.configuration.json_parser }
   let(:filter_parameters)     { HttpLog.configuration.filter_parameters }
+  let(:url_masked_body_pattern) { HttpLog.configuration.url_masked_body_pattern }
```
Comment (trusche) [inline spec/lib/http_log_spec.rb]: Align the block indentation above please.

File: lib/httplog/configuration.rb (hunk 2 above)
Comment (trusche) [reply lib/httplog/configuration.rb]: If `filter_parameters` is set, we should just mask JSON as well by default.

**Comment by Meat-Chopper** [comment]
Hello @trusche,
Thank you for review! The comments have been applied.

**Comment by trusche** [comment]
Released as v1.4.0. Thanks again!

**Pull request merged**